### PR TITLE
Go HTTP tracing examples using goBPF

### PR DESCRIPTION
**Pull request by zasgar**
This is the example for the blog post and ebpf summit talk.

### CODE REVIEW

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    u64 array_ptr_pos = buf_ptr_pos + 0;
+    u64 array_len_pos = buf_ptr_pos + 1;
+
+    //-------
+
+    // Read pointer to response struct (function's receiver, aka 'this' object).
+    void* struct_response_ptr = 0;
+    bpf_probe_read(&struct_response_ptr, sizeof(struct_response_ptr), (void *)ctx->sp+(struct_response_pos*8));
+    event.msg_len = (u64)struct_response_ptr;
+    //-------
```
Comment (oazizi000) [inline demos/simple-gotracing/http_trace/bpf_program.go:60]: nit: space above.

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    // Positions within struct request object:
+    u64 method_ptr_pos = 0;
+    u64 method_len_pos = 1;
+    u64 uri_ptr_pos = 24;
+    u64 uri_len_pos = 25;
+
+    // Positions within bufio.Writer object:
+    u64 buf_ptr_pos = 2;
+    u64 array_ptr_pos = buf_ptr_pos + 0;
+    u64 array_len_pos = buf_ptr_pos + 1;
```
Comment (oazizi000) [inline demos/simple-gotracing/http_trace/bpf_program.go:52]: Could make these const for further clarity.

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    bpf_probe_read(&array_ptr, sizeof(array_ptr), bufio_ptr+(array_ptr_pos*8));
+
+    // Read array length from buf object.
+    u64 array_len = 0;
+    bpf_probe_read(&array_len, sizeof(array_len), bufio_ptr+(array_len_pos*8));
+
+    // Read array from array pointer object.
+    u64 msg_buf_size = sizeof(event.msg);
+    msg_buf_size = msg_buf_size < content_length ? msg_buf_size : content_length;
+    // msg_buf_size = msg_buf_size < array_len      ? msg_buf_size : array_len;
```
Comment (oazizi000) [inline demos/simple-gotracing/http_trace/bpf_program.go:124]: Which version do you plan to use? Choose one and clean accordingly.

File: demos/simple-gotracing/http_trace/http_trace.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		panic(err)
+	}
+
+	// Attach the uprobe to be called everytime main.computeE is called.
+	// We need to specify the path to the binary so it can be patched.
+	// We ideally want to patch return of net/http.(*response).finishRequest, but this
+	// does not work with Go. Lookup for a fix for this soon. For now,
+	// we manually patch the correct address which is at the end of the function.
+	evName := "gobpf_uprobe"
+	err = AttachUProbeWithAddr(evName, bcc.BPF_PROBE_ENTRY, binaryProg, 0x0000000000632c61, uprobeFD, -1)
```
Comment (oazizi000) [inline demos/simple-gotracing/http_trace/http_trace.go:63]: Comment on how the address on 0x0000000000632c61was obtained.

**Review by oazizi000 — COMMENTED** [review]
Is this going to be part of the blog?

Not your fault, but the BCC code is kind of long. Might want to consider tracing less to make the code more succinct.

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    //-------
+
+    // Read pointer to response struct (function's receiver, aka 'this' object).
+    void* struct_response_ptr = 0;
+    bpf_probe_read(&struct_response_ptr, sizeof(struct_response_ptr), (void *)ctx->sp+(struct_response_pos*8));
+    event.msg_len = (u64)struct_response_ptr;
+    //-------
+
+    // Read req pointer from response struct.
+    void* req_ptr = 0;
```
Comment (yzhao1012) [inline demos/simple-gotracing/http_trace/bpf_program.go:63]: Use NULL

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    bpf_probe_read(&req_ptr, sizeof(req_ptr), struct_response_ptr+(req_ptr_pos*8));
+
+    // Read bufio pointer from response struct.
+    void* bufio_ptr = 0;
```
Comment (yzhao1012) [inline demos/simple-gotracing/http_trace/bpf_program.go:67]: NULL

File: demos/simple-gotracing/http_trace/bpf_program.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    u64 method_len_pos = 1;
+    u64 uri_ptr_pos = 24;
+    u64 uri_len_pos = 25;
+
+    // Positions within bufio.Writer object:
+    u64 buf_ptr_pos = 2;
+    u64 array_ptr_pos = buf_ptr_pos + 0;
+    u64 array_len_pos = buf_ptr_pos + 1;
+
+    //-------
```
Comment (yzhao1012) [inline demos/simple-gotracing/http_trace/bpf_program.go:54]: Not sure the intention of `//---------`

If for separating different "sections" of code, then include description of the purpose of each chunk of code could be useful.

File: demos/simple-gotracing/http_trace/http_trace.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/iovisor/gobpf/bcc"
+)
+
+/*
+#cgo CFLAGS: -I/usr/include/bcc/compat
+#cgo LDFLAGS: -lbcc
+#include <bcc/bcc_common.h>
+#include <bcc/libbpf.h>
+#include <bcc/bcc_syms.h>
+*/
```
Comment (yzhao1012) [inline demos/simple-gotracing/http_trace/http_trace.go:23]: These chunk of code need comments though.

I have no clue on what purpose they serve. Do they have special purpose outside of being plain comments?
Ignore this comment if this is well-known technique in Golang.

File: demos/simple-gotracing/http_trace/http_trace.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	Msg    [128]byte
+}
+
+var binaryProg string
+
+func init() {
+	flag.StringVar(&binaryProg, "binary", "", "The binary to probe")
+}
+
+func main() {
```
Comment (yzhao1012) [inline demos/simple-gotracing/http_trace/http_trace.go:45]: WDYT adding a short description of how to build and run this program.

I suspect this program will live longer than the blog post's attention. An instruction helps anyone who wants to play with it in the future.

**Review by oazizi000 — APPROVED** [review]

**Pull request merged**